DIPs/DIP1006.md

```diff
@@ -0,0 +1,101 @@
+and `invariant` is meant for expensive sanity checking. `in` and `out` contracts are located in between.
+
+#### Cost of contracts
+
+`assert` and `in` / `out` contracts follow a pay-for-what-you-use approach: an `assert` which is never hit is free,
+and a `in` or `out` contract on a function that is never called is also free of runtime overhead.
+
+`invariant` stand out in that regard: they can be used on a pay-for-what-you-use basis, by using `assert(object)` (or `assert(this)`),
+but they are also implicitly called twice on every `public` and `protected` class method call.
+The call itself is not direct: the compiler inserts a call to `_d_invariant`, [which can be found in druntime](https://github.com/dlang/druntime/blob/v2.072.0/src/rt/invariant.d).
+The code being very simple results in the class hierarchy to be traversed completely twice (no caching is done)
+for every call, even for classes which do not define any invariant.
+
+Experience showed that simply disabling this expensive lookup method can easily lead to 20% more throughput for OOP-intensive codebase.
+
+#### Issues with the current approach
+
+Testing can be complete and it is not rare to miss a scenario that can show up any time in a production setting.
+In such a case, having a contract not failing where it should can lead to costs higher than the one induced by a crash of the application.
+Example of such case include critical data corruption and unexpected monetary spending.
+
+For this reason, developers are often wary of disabling such safety features, even in production, as long as the performance cost is acceptable.
+What can be viewed as an acceptable performance cost is up to the end user and the code base.
+Any OOP-intensive code will want to get rid of `invariant` as a first step, while code more function-oriented might be more interested in disabling `in` / `out`.
+
+#### Considered alternatives
+
+The 4 values available to the user are voluntarily simple and hierarchical. It makes little sense to allow any contracts without enabling asserts.
+It would be feasible to allow `invariant` without `in` and `out` contracts, or only `out` contracts, only `in` contracts, or some other combination.
+Since providing all combinations would increase complexity, but doesn't provide an obvious advantage, it was left out of this proposal (but can be subject to another one).
+
+Since this proposal was heavily motivated by the cost of invariant, an obvious alternative is to reduce said cost.
+The compiler might be able to, for example, generate a virtual function which aggregates all of it's base class invariants,
+resulting in a single indirection (virtual call) when no invariant is used.
+While it's a direction that should be pursued, it is a different matter which should be pursued as a separate enhancement.
+The reason one might want to disable invariant is not because they are expensive, but because they are not a pay-for-what-you-use feature.
```

Comment:
Maybe it also worth mentioning that some compilers already offer these options (like ldc).